btree/btree.c

```diff
@@ -5814,68 +6238,318 @@ static bool vkvv_isfit(struct slot *slot)
-	    vkvv_crctype_get(node) == M0_BCT_BTREE_ENC_RAW_HASH) {
-		val_addr        = vkvv_val(slot->s_node, slot->s_idx + 1);
-		vsize           = vkvv_rec_val_size(slot->s_node, slot->s_idx);
-		calculated_csum = m0_hash_fnc_fnv1(val_addr, vsize);
+	total_size = sizeof(*dir) * (h->vkvv_dir_entries - idx - 1);
 
-		*(uint64_t*)(val_addr + vsize) = calculated_csum;
-	}
+	m0_memmove(&dir[idx], &dir[idx + 1], total_size);
+	h->vkvv_dir_entries--;
+
+	/**
+	 * Increase the size of last free fragment of value (i.e.fragment
+	 * beetween directory and last valid value).
+	 */
+	dir[h->vkvv_dir_entries - 1].val_offset     -= sizeof(*dir);
+	dir[h->vkvv_dir_entries - 1].alloc_val_size += sizeof(*dir);
 }
 
-/**
- * @brief This function will determine if the current space for values or keys
- *        overlaps with the space occupied by directory.
- */
-static bool vkvv_is_dir_overlap(struct slot *slot)
+static bool vkvv_compaction_check(struct slot *slot)
 {
-	uint32_t          ksize          =
-			m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
-	uint32_t          vsize          =
-			m0_vec_count(&slot->s_rec.r_val.ov_vec);
-	struct vkvv_head *h              = vkvv_data(slot->s_node);
-	int               count          = h->vkvv_used;
+	struct vkvv_head    *h                 = vkvv_data(slot->s_node);
+	struct vkvv_dir_rec *dir               = vkvv_dir_get(slot->s_node);
+	int                  total_avail_ksize = 0;
+	int                  total_avail_vsize = 0;
+	int                  incoming_ksize    =
+				m0_vec_count(&slot->s_rec.r_key.k_data.ov_vec);
+	int                  incoming_vsize    =
+				m0_vec_count(&slot->s_rec.r_val.ov_vec);
+	int                  i;
 
-	void             *start_dir_addr = vkvv_get_dir_addr(slot->s_node);
-	void             *key_addr       = vkvv_key(slot->s_node, count);
-	void             *val_addr       = vkvv_val(slot->s_node, count);
-	void             *end_dir_addr   = start_dir_addr +
-					   sizeof(struct dir_rec) * (count + 2);
+	if (!IS_EMBEDDED_INDIRECT(slot->s_node))
+		return false;
 
-	if (vkvv_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
-		vsize += CRC_VALUE_SIZE;
+	if (vkvv_level(slot->s_node) == 0 &&
+	    vkvv_crctype_get(slot->s_node) == M0_BCT_BTREE_ENC_RAW_HASH)
+		incoming_vsize += CRC_VALUE_SIZE;
 
-	if (key_addr + ksize > start_dir_addr ||
-	    val_addr - vsize < end_dir_addr)
+	for (i = 0; i < h->vkvv_dir_entries; i++) {
+		total_avail_ksize += (dir[i].alloc_key_size - dir[i].key_size);
+		total_avail_vsize += (dir[i].alloc_val_size - dir[i].val_size);
+	}
+
+	if ((total_avail_ksize + total_avail_vsize) >= (incoming_ksize +
+	     incoming_vsize + sizeof(*dir)))
 		return true;
 	else
 		return false;
 }
 
-/**
- * @brief This function will move directory if the current space for values or
- *        keys overlaps with the space occupied by directory.
- */
-static void vkvv_move_dir(struct slot *slot)
+static void sort_dir(struct vkvv_dir_rec *dir, int count)
```

Comment:
s/sort_dir/vkvv_sort_dir/g
change occurances in whole file.

Reply:
Done